cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/io/cdap/cdap/etl/batch/StructuredRecordComparator.java

```diff
@@ -0,0 +1,161 @@
+import java.util.Iterator;
+
+/**
+ * Compares StructuredRecords. Records cannot contain unions that do not represent a nullable type and they cannot
+ * contain maps. Arrays cannot contain non-comparable types.
+ *
+ * Records are first compared by their schemas. If their schemas are the same, they are compared field by field.
+ *
+ * Record fields are compared first by their schema fields. If the schema fields are the same, they are compared by
+ * value.
+ *
+ * Numbers, boolean, and strings are compared with their normal java compare methods. A null value is less than
+ * a non-null value.
+ *
+ * Arrays are compared by their values in order.
+ *
+ * Bytes are compared lexicographically.
+ */
+public class StructuredRecordComparator implements Comparator<StructuredRecord> {
+  private static final Comparator<Schema> SCHEMA_COMPARATOR = new SchemaComparator();
+
+  @Override
+  public int compare(StructuredRecord r1, StructuredRecord r2) {
+    if (r1 == null && r2 != null) {
+      return -1;
+    } else if (r1 != null && r2 == null) {
+      return 1;
+    } else if (r1 == null) {
+      return 0;
+    }
+
+    int comp = SCHEMA_COMPARATOR.compare(r1.getSchema(), r2.getSchema());
+    if (comp != 0) {
+      return comp;
+    }
+
+    // both records must have the same fields, otherwise their schemas would be different
+    //noinspection ConstantConditions
+    for (Schema.Field field : r1.getSchema().getFields()) {
+      comp = compareFields(field.getName(), field.getSchema(), r1.get(field.getName()), r2.get(field.getName()));
+      if (comp != 0) {
+        return comp;
+      }
+    }
+
+    return 0;
+  }
+
+  private int compareFields(String fieldName, Schema schema, Object val1, Object val2) {
+    if (val1 == null && val2 != null) {
+      return -1;
+    } else if (val1 != null && val2 == null) {
+      return 1;
+    } else if (val1 == null) {
+      return 0;
+    }
+
+    switch (schema.getType()) {
+      case NULL:
+        return 0;
+      case INT:
+        return Integer.compare((int) val1, (int) val2);
+      case LONG:
+        return Long.compare((long) val1, (long) val2);
+      case FLOAT:
+        return Float.compare((float) val1, (float) val2);
+      case DOUBLE:
+        return Double.compare((double) val1, (double) val2);
+      case STRING:
+        return ((String) val1).compareTo((String) val2);
+      case BOOLEAN:
+        return Boolean.compare((boolean) val1, (boolean) val2);
+      case ENUM:
+        return Integer.compare(((Enum) val1).ordinal(), ((Enum) val2).ordinal());
+      case BYTES:
+        return Bytes.compareTo(asBytes(fieldName, val1), asBytes(fieldName, val2));
+      case RECORD:
+        return compare((StructuredRecord) val1, (StructuredRecord) val2);
+      case ARRAY:
+        return compareArrays(fieldName, schema.getComponentSchema(), val1, val2);
+      case UNION:
+        throw new IllegalArgumentException(String.format("Field '%s' is a union. Cannot compare union fields",
```

Comment:
why can't unions be compared? Isn't any nullable field a union of null and the actual type?

Reply:
Nulls were handled specially above, but generically handling unions seemed really complicated until I saw the following from the Avro spec:

```
Unions may not contain more than one schema with the same type, except for the named types record, fixed and enum. For example, unions containing two array types or two map types are not permitted, but two types with different names are permitted. (Names permit efficient resolution when reading and writing unions.)

Unions may not immediately contain other unions.
```

I'll add support for unions, and also make sure the Schema class enforces those rules.